File: slate/ipc.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any


@dataclass(frozen=True, slots=True)
class IPCMessage:
    type: str
    payload: dict[str, Any]
    sender: str = "slate-core"

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def create_webview_ipc(tab_id: str, url: str, ad_blocked: bool) -> IPCMessage:
    return IPCMessage(
        type="BROWSER_WEBVIEW_LOAD",
        payload={
            "tab_id": tab_id,
            "url": url,
            "config": {
                "sandbox": True,
                "ad_blocked": ad_blocked,
            },
        },
    )

File: src/ipc.js
/**
 * Slate IPC message types and helpers for frontend/backend communication.
 */

/**
 * @typedef {Object} IPCMessage
 * @property {string} type
 * @property {Object} payload
 * @property {string} sender
 */

export const SENDER_CORE = 'slate-core';
export const SENDER_UI = 'slate-ui';

/**
 * @param {string} type
 * @param {Object} payload
 * @param {string} sender
 * @returns {IPCMessage}
 */
export function createIPCMessage(type, payload, sender = SENDER_CORE) {
  return {
    type,
    payload,
    sender
  };
}

export const MSG_WEBVIEW_LOAD = 'BROWSER_WEBVIEW_LOAD';

/**
 * @param {string} tabId
 * @param {string} url
 * @param {boolean} adBlocked
 * @returns {IPCMessage}
 */
export function createWebviewIPC(tabId, url, adBlocked) {
  return createIPCMessage(MSG_WEBVIEW_LOAD, {
    tabId,
    url,
    config: {
      sandbox: true,
      adBlocked
    }
  });
}
